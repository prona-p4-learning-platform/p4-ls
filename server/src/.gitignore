tree-sitter-p4.d.ts