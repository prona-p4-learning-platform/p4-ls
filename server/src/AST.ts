// take tree-sitter input and construct an ASTconst

export const ScopeNodeTypes = new Set<string>().add("");
