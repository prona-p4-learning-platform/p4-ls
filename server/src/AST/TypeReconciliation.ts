import ASTNode from "./node/ASTNode";

export function checkTypes(node1: ASTNode, node2: ASTNode) {
  throw new Error("Invalid types.");
}
