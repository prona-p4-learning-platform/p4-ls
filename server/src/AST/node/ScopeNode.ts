import { SyntaxNode } from "tree-sitter";
import { Position } from "vscode-languageserver";
import ConstantDeclarationNode from "./ConstantDeclarationNode";
import ASTNode from "./ASTNode";
import TypeDeclarationNode from "./TypeDeclarationNode";
import VariableDeclarationNode from "./VariableDeclarationNode";

export type DeclaredVariable = ConstantDeclarationNode;

export default abstract class ScopeNode extends ASTNode {
  private declaredTypes = new Map<string, TypeDeclarationNode>();
  private declaredVariables = new Map<string, DeclaredVariable>();
  private childScopeNodes: ScopeNode[] = [];
  constructor(
    private syntaxSubtree: SyntaxNode,
    public kind: string,
    private parentScopeNode: ScopeNode | null
  ) {
    super(syntaxSubtree);
  }

  addDeclaredType(type: TypeDeclarationNode) {
    this.declaredTypes.set(type.identifier, type);
  }

  addDeclaredVariable(variable: DeclaredVariable) {
    this.declaredVariables.set(variable.identifier, variable);
  }

  getDeclaredType(identifier: string) {
    return this.declaredTypes.get(identifier);
  }

  getDeclaredVariable(identifier: string) {
    return this.declaredVariables.get(identifier);
  }

  getVariableOrTypeForIdentifierAtPos(
    identifier: string,
    pos: Position
  ): TypeDeclarationNode | VariableDeclarationNode | null {
    if (this.containsPosition(pos)) {
      const c = this.childScopeNodes.filter((node) =>
        node.containsPosition(pos)
      );
      if (c.length > 0) {
        return c[0].getVariableOrTypeForIdentifierAtPos(identifier, pos);
      }
    }
    if (this.declaredVariables.has(identifier)) {
      return this.getDeclaredVariable(identifier)!;
    }
    if (this.declaredTypes.has(identifier)) {
      return this.getDeclaredType(identifier)!;
    }
    return null;
  }

  containsPosition(position: Position): boolean {
    const { startPosition, endPosition } = this.syntaxSubtree;
    if (
      startPosition.row <= position.line &&
      endPosition.row >= position.line
    ) {
      if (
        startPosition.row === position.line &&
        startPosition.column > position.character
      ) {
        return false;
      }
      if (
        endPosition.row === position.line &&
        endPosition.column < position.character
      ) {
        return false;
      }
      return true;
    }
    return false;
  }

  getParentScopeNode() {
    return this.parentScopeNode;
  }

  addChildScopeNode(child: ScopeNode) {
    this.childScopeNodes.push(child);
  }

  toJSON(): object {
    return { kind: this.kind, children: this.children };
  }
}
