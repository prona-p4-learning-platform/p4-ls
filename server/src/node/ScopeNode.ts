import { SyntaxNode } from "tree-sitter";
import { Position } from "vscode-languageserver";

type DeclaredConstantType = {
  type: "constant";
  identifier: string;
  node: SyntaxNode;
};

type DeclaredHeaderOrStructType = {
  type: "header" | "struct";
  identifier: string;
  structFields: { [key: string]: string };
  node: SyntaxNode;
};

export type DeclaredType = DeclaredHeaderOrStructType | DeclaredConstantType;

export default class ScopeNode {
  private declaredTypes = new Map<string, DeclaredType>();
  private childScopeNodes: ScopeNode[] = [];
  public type: string = "";
  constructor(
    private parentScopeNode: ScopeNode | null,
    private syntaxSubtree: SyntaxNode
  ) {}

  addDeclaredType(type: DeclaredType) {
    this.declaredTypes.set(type.identifier, type);
  }

  addDeclaredVariable(identifier: string, type: DeclaredType) {}

  containsPosition(position: Position): boolean {
    const { startPosition, endPosition } = this.syntaxSubtree;
    if (
      startPosition.row <= position.line &&
      endPosition.row >= position.line
    ) {
      if (
        startPosition.row === position.line &&
        startPosition.column > position.character
      ) {
        return false;
      }
      if (
        endPosition.row === position.line &&
        endPosition.column < position.character
      ) {
        return false;
      }
      return true;
    }
    return false;
  }

  resolveType(identifier: string): DeclaredType | null {
    if (this.declaredTypes.has(identifier)) {
      return this.declaredTypes.get(identifier) as DeclaredType;
    }
    if (this.parentScopeNode) {
      return this.parentScopeNode.resolveType(identifier);
    }
    return null;
  }

  getParentScopeNode() {
    return this.parentScopeNode;
  }

  addChildScopeNode(child: ScopeNode) {
    this.childScopeNodes.push(child);
  }

  findClosestDefinitionForIdentifier(
    identifier: string,
    position: Position
  ): SyntaxNode | null {
    let currentCandidate: SyntaxNode | null = null;
    const scopes = this.childScopeNodes.filter((node) =>
      node.containsPosition(position)
    );

    if (scopes.length > 0) {
      const resolvedType = scopes[0].resolveType(identifier);
      currentCandidate = resolvedType === null ? null : resolvedType.node;
    } else {
      const resolvedType = this.resolveType(identifier);
      if (resolvedType) {
        currentCandidate = resolvedType.node;
      }
    }

    return currentCandidate;
  }

  toJSON(): any {
    return {
      declaredTypes: [...this.declaredTypes.entries()].map((entry) => entry[0]),
      children: this.childScopeNodes.map((node) => node.toJSON()),
    };
  }
}
