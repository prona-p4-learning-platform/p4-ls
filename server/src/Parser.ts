import Parser, { SyntaxNode } from "tree-sitter";
import { Diagnostic } from "vscode-languageserver";
import ScopeNode, { DeclaredType } from "./node/ScopeNode";
import parse from "./tree-sitter-p4/parse";
import { logInfo } from "./utils/Logger"

const BlockScopeNodeTypes = new Set<string>()
  .add("blockStatement")
  .add("controlDeclaration");

const isBlockScopeNode = (type: string) => BlockScopeNodeTypes.has(type);

type ConstantDeclaration = {
  type: "constantDeclaration";
  declaredType: ASTNode;
  identifier: ASTNode;
  value: ASTNode;
};

type ScopeNode2 = {
  type: "scope";
  syntaxNode: SyntaxNode;
  statements: Array<any>;
};

type Expression = {
  type: "expression";
  value: string;
};

type Identifier = {
  type: "identifier";
  value: string;
};

type BaseType = {
  type: "basetype";
  value: string;
};

type File = {
  type: "file";
  statements: ASTNode[];
};

type ASTNode =
  | ScopeNode2
  | ConstantDeclaration
  | Expression
  | Identifier
  | BaseType
  | File;

export const createAST = (rootNode: Parser.SyntaxNode): ASTNode => {
  function recurse(tree: Parser.SyntaxNode): ASTNode {
    if (tree.type === "constantDeclaration") {
      const child0 = tree.namedChild(0);
      const child1 = tree.namedChild(1);
      const child2 = tree.namedChild(2);
      const returnVal: ConstantDeclaration = {
        type: "constantDeclaration",
        declaredType: recurse(child0!),
        identifier: recurse(child1!),
        value: recurse(child2!),
      };
      return returnVal;
    } else if (tree.type === "source_file") {
      return {
        type: "file",
        statements: tree.namedChildren.map((child) => recurse(child)),
      };
    } else if (tree.type === "baseType") {
      return {
        type: "basetype",
        value: tree.text,
      };
    } else if (tree.type === "name") {
      return {
        type: "identifier",
        value: tree.text,
      };
    } else if (tree.type === "expression") {
      return {
        type: "expression",
        value: tree.text,
      };
    }
    return tree.namedChildren.map((child) => recurse(child))[0];
  }
  return recurse(rootNode);
};

const collectTypesInScopeNodes = (rootNode: Parser.SyntaxNode): ScopeNode => {
  let currentBlockScope = new ScopeNode(null, rootNode);
  currentBlockScope.type = "root";
  function recurse(tree: Parser.SyntaxNode) {
    let updatedScope = false;
    if (isBlockScopeNode(tree.type)) {
      const oldBlockScopeNode = currentBlockScope;
      currentBlockScope = new ScopeNode(currentBlockScope, tree);
      currentBlockScope.type = tree.type;
      oldBlockScopeNode.addChildScopeNode(currentBlockScope);
      updatedScope = true;
    }

    if (tree.type === "constantDeclaration") {
      const child0 = tree.namedChild(0);
      const child1 = tree.namedChild(1);
      if (child1) {
        logInfo(child1.text + "(type: " + currentBlockScope.type + ")");
        currentBlockScope.addDeclaredType({
          type: "constant",
          identifier: child1.text,
          node: tree,
        });
      }
    } else if (tree.type === "typedefDeclaration") {
      const child1 = tree.namedChild(1);
      if (child1) {
        currentBlockScope.addDeclaredType({
          type: "constant",
          identifier: child1.text,
          node: tree,
        });
      }
    } else if (tree.type === "headerTypeDeclaration") {
      const child = tree.namedChild(0);
      const structFields = tree.descendantsOfType("structField");
      if (child) {
        const test: DeclaredType = {
          type: "header",
          identifier: child.text,
          structFields: structFields.reduce(
            (obj, node) => ({
              ...obj,
              [node.namedChild(0)!.text]: node.namedChild(1)!.text,
            }),
            {}
          ),
          node: tree,
        };
        currentBlockScope.addDeclaredType(test);
      }
    } else if (tree.type === "structTypeDeclaration") {
      const child = tree.namedChild(0);
      const structFields = tree.descendantsOfType("structField");
      if (child) {
        const test: DeclaredType = {
          type: "struct",
          identifier: child.text,
          structFields: structFields.reduce(
            (obj, node) => ({
              ...obj,
              [node.namedChild(0)!.text]: node.namedChild(1)!.text,
            }),
            {}
          ),
          node: tree,
        };
        currentBlockScope.addDeclaredType(test);
      }
    } else if (tree.type === "ERROR") {
    } else if (tree.type === "parameter") {
      const child0 = tree.namedChild(0);
      const child1 = tree.namedChild(1);
      if (child1) {
        currentBlockScope.addDeclaredType({
          type: "constant",
          identifier: child1.text,
          node: tree,
        });
      }
    }
    tree.namedChildren.forEach((n) => recurse(n));

    if (updatedScope === true) {
      const parent = currentBlockScope.getParentScopeNode();
      if (parent) {
        currentBlockScope = parent;
      }
    }
  }
  recurse(rootNode);

  return currentBlockScope;
};

export function parseSource(
  source: string
): { scopeTreeRoot: ScopeNode; parseTreeRoot: Parser.SyntaxNode } {
  const parseTreeRoot = parse(source);
  const scopeTreeRoot = collectTypesInScopeNodes(parseTreeRoot.rootNode);
  logInfo(JSON.stringify(createAST(parseTreeRoot.rootNode)));
  return {
    scopeTreeRoot,
    parseTreeRoot: parseTreeRoot.rootNode,
  };
}
