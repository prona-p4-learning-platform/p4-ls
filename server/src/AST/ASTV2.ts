import Parser from "tree-sitter";
import * as P4Parser from "../tree-sitter-p4/P4Parser";
import parse from "../tree-sitter-p4/parse";
import code from "./code";

function gotoPreorderSucc(cursor: P4Parser.TreeCursor): boolean {
  if (cursor.gotoFirstChild()) return true;
  while (!cursor.gotoNextSibling()) {
    if (!cursor.gotoParent()) {
      return false;
    }
  }
  return true;
}

const tree = parse(code) as P4Parser.Tree;

function printDeclaredNames() {
  let cursor = tree.walk();

  do {
    const c = cursor as P4Parser.TypedTreeCursor;
    switch (c.nodeType) {
      case P4Parser.SyntaxType.ConstantDeclaration: {
        let node = c.currentNode;
        break;
      }
      case P4Parser.SyntaxType.StructTypeDeclaration:
        break;
    }
  } while (gotoPreorderSucc(cursor));
}
printDeclaredNames();
