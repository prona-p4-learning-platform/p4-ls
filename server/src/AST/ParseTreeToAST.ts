import Parser from "tree-sitter";
import ActionDeclarationNode from "./node/ActionDeclarationNode";
import AssignmentStatement from "./node/AssignmentStatement";
import ASTNode from "./node/ASTNode";
import BaseTypeNode from "./node/BaseTypeNode";
import ConstantDeclarationNode from "./node/ConstantDeclarationNode";
import ControlDeclarationNode from "./node/ControlDeclarationNode";
import ExpressionNode, {
  AdditionExpression,
  DivisionExpression,
  MultiplicationExpression,
  PropertyAccessExpression,
  SubstrationExpression,
} from "./node/ExpressionNode";
import FileNode from "./node/FileNode";
import HeaderTypeDeclarationNode from "./node/HeaderTypeDeclarationNode";
import IdentifierNode from "./node/IdentifierNode";
import IntegerNode from "./node/IntegerNode";
import MethodCallStatement from "./node/MethodCallStatement";
import ParameterNode from "./node/ParameterNode";
import ParserDeclarationNode from "./node/ParserDeclarationNode";
import ParserStateNode from "./node/ParserStateNode";
import ScopeNode from "./node/ScopeNode";
import StructTypeDeclarationNode from "./node/StructTypeDeclarationNode";
import ParserStateExpression from "./node/ParserStateExpression";
import CommentNode from "./node/CommentNode";

export const createAST = (rootNode: Parser.SyntaxNode): ASTNode => {
  let currentScopeNode: ScopeNode;

  function recurse(tree: Parser.SyntaxNode): ASTNode {
    if (tree === null) {
      return new ASTNode(tree);
    }
    if (tree.type === "constantDeclaration") {
      const child0 = recurse(tree.namedChild(0)!);
      const child1 = tree.namedChild(1);
      const child2 = recurse(tree.namedChild(2)!);
      const returnVal: ConstantDeclarationNode = new ConstantDeclarationNode(
        tree,
        child0.text(),
        child1!.text,
        child2
      );
      returnVal.children = [child0, child2];
      currentScopeNode.addDeclaredVariable(returnVal);
      return returnVal;
    } else if (tree.type === "comment") {
      return new CommentNode(tree, tree.text);
    } else if (tree.type === "source_file") {
      currentScopeNode = new FileNode(tree);
      (currentScopeNode as FileNode).setStatements(
        tree.namedChildren.map((child) => recurse(child))
      );
      return currentScopeNode;
    } else if (tree.type === "baseType") {
      const node = new BaseTypeNode(
        tree,
        tree.child(0)!.text,
        tree.namedChildCount > 0 ? recurse(tree.namedChild(0)!) : null
      );
      return node;
    } else if (tree.type === "name") {
      return new IdentifierNode(tree, tree.text);
    } else if (tree.type === "stateExpression") {
      return new ParserStateExpression(tree);
    } else if (tree.type === "headerTypeDeclaration") {
      const structProps: { [key: string]: ASTNode } = {};
      tree
        .descendantsOfType("structField")
        .forEach(
          (child) =>
            (structProps[child.child(1)!.text] = recurse(child.firstChild!))
        );
      const node = new HeaderTypeDeclarationNode(
        tree,
        tree.namedChild(0)!.text,
        structProps
      );
      currentScopeNode.addDeclaredType(node);
      return node;
    } else if (tree.type === "structTypeDeclaration") {
      const structProps: { [key: string]: ASTNode } = {};
      tree
        .descendantsOfType("structField")
        .forEach(
          (child) =>
            (structProps[child.child(1)!.text] = recurse(child.firstChild!))
        );
      const node = new StructTypeDeclarationNode(
        tree,
        tree.namedChild(0)!.text,
        structProps
      );
      currentScopeNode.addDeclaredType(node);
      return node;
    } else if (tree.type === "typeName") {
      return new IdentifierNode(tree, tree.text);
    } else if (tree.type === "methodCallStatement") {
      const node = new MethodCallStatement(
        tree,
        recurse(tree.namedChild(0)!),
        tree.descendantsOfType("argument").map((child) => recurse(child))
      );
      return node;
    } else if (tree.type === "INTEGER") {
      return new IntegerNode(tree, tree.text);
    } else if (tree.type === "parameter") {
      return new ParameterNode(
        tree,
        recurse(tree.namedChild(0)!).text(),
        tree.namedChild(1)!.text
      );
    } else if (tree.type === "parserDeclaration") {
      const parameters = tree
        .namedChild(0)
        ?.descendantsOfType("parameter")
        .map((child) => recurse(child));
      console.log("params: ", parameters);
      const node = new ParserDeclarationNode(
        tree.namedChild(0)!.text,
        parameters as ParameterNode[],
        currentScopeNode,
        tree
      );
      currentScopeNode = node;
      node.setStatements(
        tree
          .descendantsOfType(["parserLocalElement", "parserState"])
          .map((child) => recurse(child))
      );
      currentScopeNode = node.getParentScopeNode()!;
      return node;
    } else if (tree.type === "parserState") {
      const [firstNamedChild, ...rest] = tree.namedChildren;
      const name = firstNamedChild.text;
      const node = new ParserStateNode(name, tree);
      node.setStatements(rest.map((child) => recurse(child)));
      return node;
    } else if (tree.type === "controlDeclaration") {
      const parameters = tree
        .namedChild(1)
        ?.descendantsOfType("parameter")
        .map((child) => recurse(child));
      console.log("params: ", parameters);
      const node = new ControlDeclarationNode(
        tree.namedChild(0)!.text,
        parameters as ParameterNode[],
        currentScopeNode,
        tree
      );
      currentScopeNode = node;
      node.setStatements(
        tree
          .descendantsOfType("controlLocalDeclaration")
          .map((child) => recurse(child))
      );
      currentScopeNode = node.getParentScopeNode()!;
      return node;
    } else if (tree.type === "lvalue") {
      const node = tree;
      if (node.childCount === 3) {
        if (node.child(1)!.text === ".") {
          const child1 = recurse(tree.namedChild(0)!);
          const child2 = recurse(tree.namedChild(1)!);
          return new PropertyAccessExpression(tree, child1, child2);
        }
        return tree.namedChildren.map((child) => recurse(child))[0];
      } else {
        return tree.namedChildren.map((child) => recurse(child))[0];
      }
    } else if (tree.type === "expression") {
      if (tree.childCount === 3) {
        const child11 = tree.namedChild(0);
        const child22 = tree.namedChild(1);
        if (child11 === null || child22 === null) {
          return new ASTNode(tree);
        }
        const child1 = recurse(child11);
        const child2 = recurse(child22);
        switch (tree.child(1)!.text) {
          case "-":
            return new SubstrationExpression(tree, child1, child2);
          case "+":
            return new AdditionExpression(tree, child1, child2);
          case ".":
            return new PropertyAccessExpression(tree, child1, child2);
          case "*":
            return new MultiplicationExpression(tree, child1, child2);
          case "/":
            return new DivisionExpression(tree, child1, child2);
        }
        return new ExpressionNode(tree, child1, child2);
      }
      return tree.namedChildren.map((child) => recurse(child))[0];
    } else if (tree.type === "IDENTIFIER") {
      return new IdentifierNode(tree, tree.text);
    } else if (tree.type === "assignmentStatement") {
      return new AssignmentStatement(
        tree,
        recurse(tree.namedChild(0)!),
        recurse(tree.namedChild(1)!)
      );
    } else if (tree.type === "actionDeclaration") {
      const node = (currentScopeNode = new ActionDeclarationNode(
        tree.namedChild(1)!.text,
        tree
          .namedChild(1)!
          .descendantsOfType("parameter")
          .map((child) => recurse(child)) as ParameterNode[],
        currentScopeNode,
        tree
      ));
      currentScopeNode.children =
        tree.lastChild?.namedChildren.map((child) => recurse(child)) || [];
      currentScopeNode = currentScopeNode.getParentScopeNode()!;
      return node;
    } else {
      console.log(tree.type);
      return tree.namedChildren.map((child) => recurse(child))[0];
    }
  }
  const tree = recurse(rootNode);
  return tree;
};
