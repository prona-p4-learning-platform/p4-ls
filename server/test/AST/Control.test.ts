import { parseSource } from "../../src/Parser";
import { createAST } from "../../src/AST/ParseTreeToAST";
import ScopeNode from "../../src/AST/node/ScopeNode";
const code = `
control MyIngress(inout headers hdr,
    inout metadata meta,
    inout standard_metadata_t standard_metadata) {
        
    action drop() {
        mark_to_drop(standard_metadata);
    }

    action ipv4_forward(macAddr_t dstAddr, egressSpec_t port) {
        standard_metadata.egress_spec = port;
        hdr.ethernet.srcAddr = hdr.ethernet.dstAddr;
        hdr.ethernet.dstAddr = dstAddr;
        hdr.ipv4.ttl = hdr.ipv4.ttl - 1;
    }

    apply {
        if (hdr.ipv4.isValid()) {
            ipv4_lpm.apply();
        }
    }    
}
`;

it("creates a control AST", () => {
  const root = createAST(parseSource(code).parseTreeRoot);
  const type = (root as ScopeNode).getDeclaredType("ipv4_t");
  expect(root).toMatchSnapshot();
});
