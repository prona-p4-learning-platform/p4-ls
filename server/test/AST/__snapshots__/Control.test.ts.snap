// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a control AST 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "callee": Object {
                "children": Array [],
                "endPosition": Object {
                  "character": 20,
                  "line": 6,
                },
                "identifier": "mark_to_drop",
                "kind": "Identifier",
                "startPosition": Object {
                  "character": 8,
                  "line": 6,
                },
              },
              "children": Array [],
              "endPosition": Object {
                "character": 40,
                "line": 6,
              },
              "kind": "MethodCallStatement",
              "parameters": Array [
                Object {
                  "children": Array [],
                  "endPosition": Object {
                    "character": 38,
                    "line": 6,
                  },
                  "identifier": "standard_metadata",
                  "kind": "Identifier",
                  "startPosition": Object {
                    "character": 21,
                    "line": 6,
                  },
                },
              ],
              "startPosition": Object {
                "character": 8,
                "line": 6,
              },
            },
          ],
          "kind": "ActionDeclarationNode",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "endPosition": Object {
                "character": 45,
                "line": 10,
              },
              "kind": "AssignmentStatement",
              "lhs": Object {
                "children": Array [],
                "endPosition": Object {
                  "character": 37,
                  "line": 10,
                },
                "kind": "PropertyAccessExpression",
                "lhs": Object {
                  "children": Array [],
                  "endPosition": Object {
                    "character": 25,
                    "line": 10,
                  },
                  "identifier": "standard_metadata",
                  "kind": "Identifier",
                  "startPosition": Object {
                    "character": 8,
                    "line": 10,
                  },
                },
                "rhs": Object {
                  "children": Array [],
                  "endPosition": Object {
                    "character": 37,
                    "line": 10,
                  },
                  "identifier": "egress_spec",
                  "kind": "Identifier",
                  "startPosition": Object {
                    "character": 26,
                    "line": 10,
                  },
                },
                "startPosition": Object {
                  "character": 8,
                  "line": 10,
                },
              },
              "rhs": Object {
                "children": Array [],
                "endPosition": Object {
                  "character": 44,
                  "line": 10,
                },
                "identifier": "port",
                "kind": "Identifier",
                "startPosition": Object {
                  "character": 40,
                  "line": 10,
                },
              },
              "startPosition": Object {
                "character": 8,
                "line": 10,
              },
            },
            Object {
              "children": Array [],
              "endPosition": Object {
                "character": 52,
                "line": 11,
              },
              "kind": "AssignmentStatement",
              "lhs": Object {
                "children": Array [],
                "endPosition": Object {
                  "character": 28,
                  "line": 11,
                },
                "kind": "PropertyAccessExpression",
                "lhs": Object {
                  "children": Array [],
                  "endPosition": Object {
                    "character": 20,
                    "line": 11,
                  },
                  "kind": "PropertyAccessExpression",
                  "lhs": Object {
                    "children": Array [],
                    "endPosition": Object {
                      "character": 11,
                      "line": 11,
                    },
                    "identifier": "hdr",
                    "kind": "Identifier",
                    "startPosition": Object {
                      "character": 8,
                      "line": 11,
                    },
                  },
                  "rhs": Object {
                    "children": Array [],
                    "endPosition": Object {
                      "character": 20,
                      "line": 11,
                    },
                    "identifier": "ethernet",
                    "kind": "Identifier",
                    "startPosition": Object {
                      "character": 12,
                      "line": 11,
                    },
                  },
                  "startPosition": Object {
                    "character": 8,
                    "line": 11,
                  },
                },
                "rhs": Object {
                  "children": Array [],
                  "endPosition": Object {
                    "character": 28,
                    "line": 11,
                  },
                  "identifier": "srcAddr",
                  "kind": "Identifier",
                  "startPosition": Object {
                    "character": 21,
                    "line": 11,
                  },
                },
                "startPosition": Object {
                  "character": 8,
                  "line": 11,
                },
              },
              "rhs": Object {
                "children": Array [],
                "endPosition": Object {
                  "character": 51,
                  "line": 11,
                },
                "kind": "PropertyAccessExpression",
                "lhs": Object {
                  "children": Array [],
                  "endPosition": Object {
                    "character": 43,
                    "line": 11,
                  },
                  "kind": "PropertyAccessExpression",
                  "lhs": Object {
                    "children": Array [],
                    "endPosition": Object {
                      "character": 34,
                      "line": 11,
                    },
                    "identifier": "hdr",
                    "kind": "Identifier",
                    "startPosition": Object {
                      "character": 31,
                      "line": 11,
                    },
                  },
                  "rhs": Object {
                    "children": Array [],
                    "endPosition": Object {
                      "character": 43,
                      "line": 11,
                    },
                    "identifier": "ethernet",
                    "kind": "Identifier",
                    "startPosition": Object {
                      "character": 35,
                      "line": 11,
                    },
                  },
                  "startPosition": Object {
                    "character": 31,
                    "line": 11,
                  },
                },
                "rhs": Object {
                  "children": Array [],
                  "endPosition": Object {
                    "character": 51,
                    "line": 11,
                  },
                  "identifier": "dstAddr",
                  "kind": "Identifier",
                  "startPosition": Object {
                    "character": 44,
                    "line": 11,
                  },
                },
                "startPosition": Object {
                  "character": 31,
                  "line": 11,
                },
              },
              "startPosition": Object {
                "character": 8,
                "line": 11,
              },
            },
            Object {
              "children": Array [],
              "endPosition": Object {
                "character": 39,
                "line": 12,
              },
              "kind": "AssignmentStatement",
              "lhs": Object {
                "children": Array [],
                "endPosition": Object {
                  "character": 28,
                  "line": 12,
                },
                "kind": "PropertyAccessExpression",
                "lhs": Object {
                  "children": Array [],
                  "endPosition": Object {
                    "character": 20,
                    "line": 12,
                  },
                  "kind": "PropertyAccessExpression",
                  "lhs": Object {
                    "children": Array [],
                    "endPosition": Object {
                      "character": 11,
                      "line": 12,
                    },
                    "identifier": "hdr",
                    "kind": "Identifier",
                    "startPosition": Object {
                      "character": 8,
                      "line": 12,
                    },
                  },
                  "rhs": Object {
                    "children": Array [],
                    "endPosition": Object {
                      "character": 20,
                      "line": 12,
                    },
                    "identifier": "ethernet",
                    "kind": "Identifier",
                    "startPosition": Object {
                      "character": 12,
                      "line": 12,
                    },
                  },
                  "startPosition": Object {
                    "character": 8,
                    "line": 12,
                  },
                },
                "rhs": Object {
                  "children": Array [],
                  "endPosition": Object {
                    "character": 28,
                    "line": 12,
                  },
                  "identifier": "dstAddr",
                  "kind": "Identifier",
                  "startPosition": Object {
                    "character": 21,
                    "line": 12,
                  },
                },
                "startPosition": Object {
                  "character": 8,
                  "line": 12,
                },
              },
              "rhs": Object {
                "children": Array [],
                "endPosition": Object {
                  "character": 38,
                  "line": 12,
                },
                "identifier": "dstAddr",
                "kind": "Identifier",
                "startPosition": Object {
                  "character": 31,
                  "line": 12,
                },
              },
              "startPosition": Object {
                "character": 8,
                "line": 12,
              },
            },
            Object {
              "children": Array [],
              "endPosition": Object {
                "character": 40,
                "line": 13,
              },
              "kind": "AssignmentStatement",
              "lhs": Object {
                "children": Array [],
                "endPosition": Object {
                  "character": 20,
                  "line": 13,
                },
                "kind": "PropertyAccessExpression",
                "lhs": Object {
                  "children": Array [],
                  "endPosition": Object {
                    "character": 16,
                    "line": 13,
                  },
                  "kind": "PropertyAccessExpression",
                  "lhs": Object {
                    "children": Array [],
                    "endPosition": Object {
                      "character": 11,
                      "line": 13,
                    },
                    "identifier": "hdr",
                    "kind": "Identifier",
                    "startPosition": Object {
                      "character": 8,
                      "line": 13,
                    },
                  },
                  "rhs": Object {
                    "children": Array [],
                    "endPosition": Object {
                      "character": 16,
                      "line": 13,
                    },
                    "identifier": "ipv4",
                    "kind": "Identifier",
                    "startPosition": Object {
                      "character": 12,
                      "line": 13,
                    },
                  },
                  "startPosition": Object {
                    "character": 8,
                    "line": 13,
                  },
                },
                "rhs": Object {
                  "children": Array [],
                  "endPosition": Object {
                    "character": 20,
                    "line": 13,
                  },
                  "identifier": "ttl",
                  "kind": "Identifier",
                  "startPosition": Object {
                    "character": 17,
                    "line": 13,
                  },
                },
                "startPosition": Object {
                  "character": 8,
                  "line": 13,
                },
              },
              "rhs": Object {
                "children": Array [],
                "endPosition": Object {
                  "character": 39,
                  "line": 13,
                },
                "kind": "SubstrationExpression",
                "lhs": Object {
                  "children": Array [],
                  "endPosition": Object {
                    "character": 35,
                    "line": 13,
                  },
                  "kind": "PropertyAccessExpression",
                  "lhs": Object {
                    "children": Array [],
                    "endPosition": Object {
                      "character": 31,
                      "line": 13,
                    },
                    "kind": "PropertyAccessExpression",
                    "lhs": Object {
                      "children": Array [],
                      "endPosition": Object {
                        "character": 26,
                        "line": 13,
                      },
                      "identifier": "hdr",
                      "kind": "Identifier",
                      "startPosition": Object {
                        "character": 23,
                        "line": 13,
                      },
                    },
                    "rhs": Object {
                      "children": Array [],
                      "endPosition": Object {
                        "character": 31,
                        "line": 13,
                      },
                      "identifier": "ipv4",
                      "kind": "Identifier",
                      "startPosition": Object {
                        "character": 27,
                        "line": 13,
                      },
                    },
                    "startPosition": Object {
                      "character": 23,
                      "line": 13,
                    },
                  },
                  "rhs": Object {
                    "children": Array [],
                    "endPosition": Object {
                      "character": 35,
                      "line": 13,
                    },
                    "identifier": "ttl",
                    "kind": "Identifier",
                    "startPosition": Object {
                      "character": 32,
                      "line": 13,
                    },
                  },
                  "startPosition": Object {
                    "character": 23,
                    "line": 13,
                  },
                },
                "rhs": Object {
                  "children": Array [],
                  "endPosition": Object {
                    "character": 39,
                    "line": 13,
                  },
                  "kind": "Integer",
                  "startPosition": Object {
                    "character": 38,
                    "line": 13,
                  },
                  "value": "1",
                },
                "startPosition": Object {
                  "character": 23,
                  "line": 13,
                },
              },
              "startPosition": Object {
                "character": 8,
                "line": 13,
              },
            },
          ],
          "kind": "ActionDeclarationNode",
        },
      ],
      "kind": "ControlDeclaration",
      "parameters": Array [
        Object {
          "children": Array [],
          "endPosition": Object {
            "character": 35,
            "line": 1,
          },
          "identifier": "hdr",
          "kind": "ParameterNode",
          "startPosition": Object {
            "character": 18,
            "line": 1,
          },
          "type": "headers",
        },
        Object {
          "children": Array [],
          "endPosition": Object {
            "character": 23,
            "line": 2,
          },
          "identifier": "meta",
          "kind": "ParameterNode",
          "startPosition": Object {
            "character": 4,
            "line": 2,
          },
          "type": "metadata",
        },
        Object {
          "children": Array [],
          "endPosition": Object {
            "character": 47,
            "line": 3,
          },
          "identifier": "standard_metadata",
          "kind": "ParameterNode",
          "startPosition": Object {
            "character": 4,
            "line": 3,
          },
          "type": "standard_metadata_t",
        },
      ],
    },
  ],
  "kind": "File",
}
`;
